plonky2x/src/backend/circuit/serialization.rs

```diff
@@ -0,0 +1,423 @@
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("SerializationRegistry")
+            .field("ids of registered objects", &self.registry.keys())
+            .field("index", &self.index)
+            .field("type_ids", &self.type_ids)
+            .field("current_index", &self.current_index)
+            .finish()
+    }
+}
+
+impl<F: RichField + Extendable<D>, K: Hash, T: Any, const D: usize>
+    SerializationRegistry<K, F, T, D>
+{
+    pub fn new() -> Self {
+        Self {
+            registry: HashMap::new(),
+            index: HashMap::new(),
+            type_ids: Vec::new(),
+            current_index: 0,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct WitnessGeneratorSerializerFn<W>(PhantomData<W>);
+
+#[derive(Clone)]
+pub struct GateSerializerFn<G>(PhantomData<G>);
+
+impl<F: RichField + Extendable<D>, W: WitnessGenerator<F, D>, const D: usize>
+    Serializer<F, WitnessGeneratorRef<F, D>, D> for WitnessGeneratorSerializerFn<W>
+{
+    fn read(
+        &self,
+        buf: &mut Buffer,
+        common_data: &CommonCircuitData<F, D>,
+    ) -> IoResult<WitnessGeneratorRef<F, D>> {
+        let generator: IoResult<W> = WitnessGenerator::<F, D>::deserialize(buf, common_data);
+        generator.map(|g| WitnessGeneratorRef::<F, D>::new(g))
+    }
+
+    fn write(
+        &self,
+        buf: &mut Vec<u8>,
+        object: &WitnessGeneratorRef<F, D>,
+        common_data: &CommonCircuitData<F, D>,
+    ) -> IoResult<()> {
+        object.0.serialize(buf, common_data)
+    }
+}
+
+impl<F: RichField + Extendable<D>, G: AnyGate<F, D>, const D: usize> Serializer<F, GateRef<F, D>, D>
+    for GateSerializerFn<G>
+{
+    fn read(
+        &self,
+        buf: &mut Buffer,
+        common_data: &CommonCircuitData<F, D>,
+    ) -> IoResult<GateRef<F, D>> {
+        let gate: IoResult<G> = Gate::<F, D>::deserialize(buf, common_data);
+        gate.map(|g| GateRef::<F, D>::new(g))
+    }
+
+    fn write(
+        &self,
+        buf: &mut Vec<u8>,
+        object: &GateRef<F, D>,
+        common_data: &CommonCircuitData<F, D>,
+    ) -> IoResult<()> {
+        object.0.serialize(buf, common_data)
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> WitnessGeneratorRegistry<F, D> {
+    /// Registers a new witness generator with the given id.
+    pub fn register<W: WitnessGenerator<F, D>>(&mut self, id: String) {
+        let exists = self.0.registry.insert(
+            id.clone(),
+            Box::new(WitnessGeneratorSerializerFn::<W>(PhantomData)),
+        );
+
+        if exists.is_some() {
+            panic!("Generator type {} already registered", id);
+        }
+
+        self.0.type_ids.push(id.clone());
+        self.0.index.insert(id, self.0.current_index);
+        self.0.current_index += 1;
+    }
+
+    /// Registers a new simple witness generator with the given id.
+    pub fn register_simple_generator<SG: SimpleGenerator<F, D>>(&mut self, id: String) {
```

Comment:
Rename register_simple_generator to register_simple to make the code below more readable?